README.rst
django-s3-storage
=================

**django-s3-storage** provides a Django Amazon S3 file storage.


Features
--------

- Django file storage for Amazon S3.
- Django static file storage for Amazon S3.
- Works in Python 3!


Installation
------------

1. Install using ``pip install django-s3-storage``.
2. Add ``'django_s3_storage'`` to your ``INSTALLED_APPS`` setting.
3. Set your ``DEFAULT_FILE_STORAGE`` setting to ``"django_s3_storage.storage.S3Storage"``.
4. Set your ``STATICFILES_STORAGE`` setting to ``"django_s3_storage.storage.StaticS3Storage"`` or ``"django_s3_storage.storage.ManifestStaticS3Storage"``.
5. Configure your Amazon S3 settings (see Available settings, below).


Authentication settings
-----------------------

Use the following settings to authenticate with Amazon AWS.

.. code:: python

    # The AWS region to connect to.
    AWS_REGION = "us-east-1"

    # The AWS access key to use.
    AWS_ACCESS_KEY_ID = ""

    # The AWS secret access key to use.
    AWS_SECRET_ACCESS_KEY = ""

    # The optional AWS session token to use.
    AWS_SESSION_TOKEN = ""


File storage settings
---------------------

Use the following settings to configure the S3 file storage. You must provide at least ``AWS_S3_BUCKET_NAME``.

.. code:: python

    # The name of the bucket to store files in.
    AWS_S3_BUCKET_NAME = ""

    # How to construct S3 URLs ("auto", "path", "virtual").
    AWS_S3_ADDRESSING_STYLE = "auto"

    # The full URL to the S3 endpoint. Leave blank to use the default region URL.
    AWS_S3_ENDPOINT_URL = ""

    # A prefix to be applied to every stored file. This will be joined to every filename using the "/" separator.
    AWS_S3_KEY_PREFIX = ""

    # Whether to enable authentication for stored files. If True, then generated URLs will include an authentication
    # token valid for `AWS_S3_MAX_AGE_SECONDS`. If False, then generated URLs will not include an authentication token,
    # and their permissions will be set to "public-read".
    AWS_S3_BUCKET_AUTH = True

    # How long generated URLs are valid for. This affects the expiry of authentication tokens if `AWS_S3_BUCKET_AUTH`
    # is True. It also affects the "Cache-Control" header of the files.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_MAX_AGE_SECONDS = 60 * 60  # 1 hours.

    # A URL prefix to be used for generated URLs. This is useful if your bucket is served through a CDN.
    AWS_S3_PUBLIC_URL = ""

    # If True, then files will be stored with reduced redundancy. Check the S3 documentation and make sure you
    # understand the consequences before enabling.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_REDUCED_REDUNDANCY = False

    # The Content-Disposition header used when the file is downloaded. This can be a string, or a function taking a
    # single `name` argument.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_CONTENT_DISPOSITION = ""

    # The Content-Language header used when the file is downloaded. This can be a string, or a function taking a
    # single `name` argument.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_CONTENT_LANGUAGE = ""

    # A mapping of custom metadata for each file. Each value can be a string, or a function taking a
    # single `name` argument.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_METADATA = {}

    # If True, then files will be stored using AES256 server-side encryption.
    # If this is a string value (e.g., "aws:kms"), that encryption type will be used.
    # Otherwise, server-side encryption is not be enabled.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_ENCRYPT_KEY = False

    # The AWS S3 KMS encryption key ID (the `SSEKMSKeyId` parameter) is set from this string if present.
    # This is only relevant if AWS S3 KMS server-side encryption is enabled (above).
    AWS_S3_KMS_ENCRYPTION_KEY_ID = ""

    # If True, then text files will be stored using gzip content encoding. Files will only be gzipped if their
    # compressed size is smaller than their uncompressed size.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_GZIP = True

    # The signature version to use for S3 requests.
    AWS_S3_SIGNATURE_VERSION = None

    # If True, then files with the same name will overwrite each other. By default it's set to False to have
    # extra characters appended.
    AWS_S3_FILE_OVERWRITE = False

    # If True, use default behaviour for boto3 of using threads when doing S3 operations. If gevent or similar
    # is used it must be disabled
    AWS_S3_USE_THREADS = True

    # Max pool of connections for massive S3 interactions
    AWS_S3_MAX_POOL_CONNECTIONS = 10

    # Time to raise timeout when submitting a new file
    AWS_S3_CONNECT_TIMEOUT = 60

**Important:** Several of these settings (noted above) will not affect existing files. To sync the new settings to
existing files, run ``./manage.py s3_sync_meta django.core.files.storage.default_storage``.

These settings can be provided in field storage definition like this:

.. code:: python

    from django.db import models

    from django_s3_storage.storage import S3Storage

    storage = S3Storage(aws_s3_bucket_name='test_bucket')


    class Car(models.Model):
        name = models.CharField(max_length=255)
        photo = models.ImageField(storage=storage)

**Note:** settings key in storage definition should be `lowercase`.

Staticfiles storage settings
----------------------------

All of the file storage settings are available for the staticfiles storage, suffixed with ``_STATIC``. You must provide
at least ``AWS_S3_BUCKET_NAME_STATIC``. Remember to run ``./manage.py collectstatic`` after changing your staticfiles
storage backend.

The following staticfiles storage settings have different default values to their file storage counterparts.

.. code:: python

    AWS_S3_BUCKET_AUTH_STATIC = False


The following additional staticfiles storage settings also exist:

.. code:: python

    # For ManifestStaticS3Storage, how long the browser should cache md5-hashed filenames.  This affects the expiry of
    # authentication tokens if `AWS_S3_BUCKET_AUTH` is True. It also affects the "Cache-Control" header of the files.
    # Important: Changing this setting will not affect existing files.
    AWS_S3_MAX_AGE_SECONDS_CACHED_STATIC = 60 * 60 * 24 * 365  # 1 year.


**Important:** Several of these settings (noted above) will not affect existing files. To sync the new settings to
existing files, run ``./manage.py s3_sync_meta django.contrib.staticfiles.storage.staticfiles_storage``.


Optimizing media file caching
-----------------------------

The default settings assume that media file are private. This means that they are only accessible via S3 authenticated URLs, which is bad for browser caching.

To make media files public, and enable aggressive caching, make the following changes to your ``settings.py``.

.. code:: python

    AWS_S3_BUCKET_AUTH = False

    AWS_S3_MAX_AGE_SECONDS = 60 * 60 * 24 * 365  # 1 year.

**Important:** By making these changes, all media files will be public. Ensure they do not contain confidential information.

The default settings for staticfiles storage are already optimizing for aggressive caching.


Custom URLs
-----------

Sometimes the default settings aren't flexible enough and custom handling of object is needed. For
example, the ``Content-Disposition`` might be set to force download of a file instead of opening
it:

.. code:: python

    url = storage.url("foo/bar.pdf", extra_params={"ResponseContentDisposition": "attachment"})

Another example is a link to a specific version of the file (within a bucket that has versioning
enabled):

.. code:: python

    url = storage.url("foo/bar.pdf", extra_params={"VersionId": "FRy3fTduRtqHsRAoNp0REzPJj_WunDfl"})

The ``extra_params`` dict accepts the same parameters as `get_object() <https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_object>`_.

Please note, however, that **custom URLs will not work with AWS_S3_PUBLIC_URL** where the
URL doesn't accept extra parameters, and it will raise ``ValueError``.


Pre-signed URL uploads
----------------------

Pre-signed URLs allow temporary access to S3 objects without AWS credentials. A pre-signed URL allows HTTP clients to
upload files directly, improving performance and reducing the load on your server.

To generate a presigned URL allowing a file upload with HTTP ``PUT``:

.. code:: python

    url = storage.url("foo/bar.pdf", client_method="put_object")


Management commands
-------------------

``s3_sync_meta``
~~~~~~~~~~~~~~~~

Syncronizes the meta information on S3 files.

Several settings (noted above) will not affect existing files. Run this command to sync the new settings to existing files.

Example usage: ``./manage.py s3_sync_meta django.core.files.storage.default_storage``


IAM permissions

CHANGELOG.rst
django-s3-storage changelog
===========================

0.15.0
------

- Added support for opening files in text mode on Python 3.11+ (@etianen).

0.14.0
------

- Added ``client_method`` to ``save()`` method, allowing the generation of presigned URLs for direct S3 upload (@Roriz).


0.13.11
-------

- Added ``AWS_S3_CONNECT_TIMEOUT`` setting (@roriz).

0.13.10
-------

- Added ``AWS_S3_MAX_POOL_CONNECTIONS`` setting (@Routhinator).


0.13.9
------

- Alternative fix for ``ValueError`` when using ``ManifestStaticS3Storage`` (@etianen).

0.13.8
------

- Fixed ``RemovedInDjango50Warning`` warning from use of ``django.utils.timezone.utc`` (@kevinmarsh).

0.13.7
------

- Fixed ``ValueError`` when using ``ManifestStaticS3Storage`` (@etianen).


0.13.6
------

- Adding `AWS_S3_USE_THREADS` to fix `gevent` issues (@uxio0).
- Add `copy()` and `rename()` methods (@emesik).


0.13.5
------

- Saving files now uses multipart streaming upload, supporting arbitrarily large files without buffering contents to
  memory (@bramverhelst).
- Fixed some deprecation warnings (@kevinmarsh).
- Documentation fixes (@riggedCoinflip, @dduong42).
- CI improvements (@michael-k, @etianen).


0.13.4
------

- Allowing both `AWS_S3_BUCKET_AUTH` and `AWS_S3_PUBLIC_URL` to be set (with a stern warning!) (@ipmb).


0.13.3
------

- Fixed error in `exists()` when using DigitalOcean's AMS3 (@heckad).


0.13.2
------

- Removed validation of `AWS_SECRET_ACCESS_KEY`, `AWS_ACCESS_KEY_ID`, and `AWS_SESSION_TOKEN`, because these can
  be detected by boto by using the AWS instance metadata service (@MichaelAnckaert).


0.13.1
------

- Added further settings validation to help newcomers to the library (@etianen).


0.13.0
------

- Django 3.0 support (@michael-k, @etianen).
- Removed `S3Error`, replaced with built-in `OSError` and subclasses (@etianen).


0.12.5
------

- Added support for custom URL parameters in calls to `url()` (@emesik).
- Calls to `size()` now report uncompressed size when `AWS_S3_GZIP` is enabled (@emesik).
- Allowing storage classes to be picked (@fdemmer).
- Documentation improvements, bugfixes and minor tweaks (@hramezani, @michael-k).


0.12.4
------

- Added support for AWS KMS encryption, using the ``AWS_S3_ENCRYPT_KEY`` and ``AWS_S3_KMS_ENCRYPTION_KEY_ID`` settings
  (@MartinFalatic).


0.12.3
------

- Actually fixed issues using ``S3Storage`` in a multithreaded environment.


0.12.2
------

- Fixed issues using ``S3Storage`` in a multithreaded environment.


0.12.1
------

- Fixed 'header does not match what was computed' error (@etianen).
- Compatibility with riak-cs (@flo-dhalluin).
- Fixed documentation typos (@mgalgs).


0.12.0
------

- Changed default for `AWS_S3_MAX_AGE_SECONDS_STATIC` to 1 hour (see https://github.com/etianen/django-s3-storage/issues/62) (@etianen, @marfire).
- Added `AWS_S3_MAX_AGE_SECONDS_CACHED_STATIC` setting (@etianen, @marfire).


0.11.3
------

- Added ``AWS_S3_SIGNATURE_VERSION`` setting.
- Changed the default signature version for S3 to v4.
  According to the `AWS documentation <http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region>`_ all S3 regions support v4 (but not all support v2).
- Raising ``S3Error`` instead of ``OSError`` if S3 storage throws an error. ``S3Error`` inherits from both ``OSError`` and ``IOError``.
- Better checking for directory existance (@kencochrane, @etianen).
- Added ``AWS_S3_FILE_OVERWRITE `` setting (@Edke).


0.11.2
------

- Bugfix: Fixed ``listdir()`` at bucket root returning an empty list (@aaugustin).
- Added ``get_modified_time`` support (@jschneier).
- Testing against Django 1.11 (@matthiask).


0.11.1
------

- Raising ``OSError`` instead of ``IOError`` if S3 storage throws an error. On Python 3 it makes no difference, but on Python 2 it's what collectstatic expects.
- Fixed issue with ``s3_sync_meta`` where a race condition or key name normalization could cause an ``OSError`` to be raised.
- Fixed `modified_time()` on non-UTC machines.


0.11.0
------

- *Breaking:* All S3 keys are normalized to use unix-style path separators, and resolve relative paths.


0.10.0
------

- Switched to `boto3`-based implementation.
- Added `AWS_S3_CONTENT_DISPOSITION` and `AWS_S3_CONTENT_LANGUAGE` settings.
- Added `AWS_S3_CONTENT_DISPOSITION_STATIC` and `AWS_S3_CONTENT_LANGUAGE_STATIC` settings.
- *Breaking:* Setting Content-Disposition and Content-Language headers via `AWS_S3_METADATA` setting no longer supported.
- *Breaking:* `AWS_S3_HOST` setting refactored to `AWS_S3_ENDPOINT_URL`.

django_s3_storage/storage.py
import gzip
import logging
import mimetypes
import os
import posixpath
import shutil
from contextlib import closing
from datetime import timezone
from functools import partial, wraps
from io import TextIOBase, TextIOWrapper
from tempfile import SpooledTemporaryFile
from threading import local
from urllib.parse import urljoin, urlsplit, urlunsplit

import boto3
from boto3.s3.transfer import TransferConfig
from botocore.client import Config
from botocore.exceptions import ClientError
from django.conf import settings
from django.contrib.staticfiles.storage import ManifestFilesMixin
from django.core import checks
from django.core.exceptions import ImproperlyConfigured
from django.core.files.base import File
from django.core.files.storage import Storage
from django.core.signals import setting_changed
from django.utils.deconstruct import deconstructible
from django.utils.encoding import filepath_to_uri, force_bytes, force_str
from django.utils.timezone import make_naive

log = logging.getLogger(__name__)


def _wrap_errors(func):
    @wraps(func)
    def _do_wrap_errors(self, name, *args, **kwargs):
        try:
            return func(self, name, *args, **kwargs)
        except ClientError as ex:
            code = ex.response.get("Error", {}).get("Code", "Unknown")
            err_cls = OSError
            if code == "NoSuchKey":
                err_cls = FileNotFoundError
            raise err_cls(f"S3Storage error at {name!r}: {force_str(ex)}")

    return _do_wrap_errors


def _callable_setting(value, name):
    return value(name) if callable(value) else value


def _to_sys_path(name):
    return name.replace("/", os.sep)


def _to_posix_path(name):
    return name.replace(os.sep, "/")


def _wrap_path_impl(func):
    @wraps(func)
    def do_wrap_path_impl(self, name, *args, **kwargs):
        # The default implementations of most storage methods assume that system-native paths are used. But we deal with
        # posix paths. We fix this by converting paths to system form, passing them to the default implementation, then
        # converting them back to posix paths.
        return _to_posix_path(func(self, _to_sys_path(name), *args, **kwargs))

    return do_wrap_path_impl


def unpickle_helper(cls, kwargs):
    return cls(**kwargs)


Settings = type(force_str("Settings"), (), {})


_UNCOMPRESSED_SIZE_META_KEY = "uncompressed_size"


class S3File(File):

    """
    A file returned from Amazon S3.
    """

    def __init__(self, file, name, storage):
        super().__init__(file, name)
        self._storage = storage

    def open(self, mode="rb"):
        if self.closed:
            self.file = self._storage.open(self.name, mode).file
        return super().open(mode)


class _Local(local):

    """
    Thread-local connection manager.

    Boto3 objects are not thread-safe.
    https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html#multithreading-and-multiprocessing
    """

    def __init__(self, storage):
        connection_kwargs = {
            "region_name": storage.settings.AWS_REGION,
        }
        if storage.settings.AWS_ACCESS_KEY_ID:
            connection_kwargs["aws_access_key_id"] = storage.settings.AWS_ACCESS_KEY_ID
        if storage.settings.AWS_SECRET_ACCESS_KEY:
            connection_kwargs[
                "aws_secret_access_key"
            ] = storage.settings.AWS_SECRET_ACCESS_KEY
        if storage.settings.AWS_SESSION_TOKEN:
            connection_kwargs["aws_session_token"] = storage.settings.AWS_SESSION_TOKEN
        if storage.settings.AWS_S3_ENDPOINT_URL:
            connection_kwargs["endpoint_url"] = storage.settings.AWS_S3_ENDPOINT_URL
        self.session = boto3.session.Session()
        self.s3_connection = self.session.client(
            "s3",
            config=Config(
                s3={"addressing_style": storage.settings.AWS_S3_ADDRESSING_STYLE},
                signature_version=storage.settings.AWS_S3_SIGNATURE_VERSION,
                max_pool_connections=storage.settings.AWS_S3_MAX_POOL_CONNECTIONS,
                connect_timeout=storage.settings.AWS_S3_CONNECT_TIMEOUT,
            ),
            **connection_kwargs,
        )


@deconstructible
class S3Storage(Storage):

    """
    An implementation of Django file storage over S3.
    """

    default_auth_settings = {
        "AWS_REGION": "us-east-1",
        "AWS_ACCESS_KEY_ID": "",
        "AWS_SECRET_ACCESS_KEY": "",
        "AWS_SESSION_TOKEN": "",
    }

    default_s3_settings = {
        "AWS_S3_BUCKET_NAME": "",
        "AWS_S3_ADDRESSING_STYLE": "auto",
        "AWS_S3_ENDPOINT_URL": "",
        "AWS_S3_KEY_PREFIX": "",
        "AWS_S3_BUCKET_AUTH": True,
        "AWS_S3_MAX_AGE_SECONDS": 60 * 60,  # 1 hours.
        "AWS_S3_PUBLIC_URL": "",
        "AWS_S3_REDUCED_REDUNDANCY": False,
        "AWS_S3_CONTENT_DISPOSITION": "",
        "AWS_S3_CONTENT_LANGUAGE": "",
        "AWS_S3_METADATA": {},
        "AWS_S3_ENCRYPT_KEY": False,
        "AWS_S3_KMS_ENCRYPTION_KEY_ID": "",
        "AWS_S3_GZIP": True,
        "AWS_S3_SIGNATURE_VERSION": "s3v4",
        "AWS_S3_FILE_OVERWRITE": False,
        "AWS_S3_USE_THREADS": True,
        "AWS_S3_MAX_POOL_CONNECTIONS": 10,
        "AWS_S3_CONNECT_TIMEOUT": 60,  # 60 seconds
    }

    s3_settings_suffix = ""

    def _setup(self):
        self.settings = Settings()
        # Configure own settings.
        for setting_key, setting_default_value in self.default_auth_settings.items():
            setattr(
                self.settings,
                setting_key,
                self._kwargs.get(
                    setting_key.lower(),
                    getattr(settings, setting_key, setting_default_value),
                ),
            )
        for setting_key, setting_default_value in self.default_s3_settings.items():
            setattr(
                self.settings,
                setting_key,
                self._kwargs.get(
                    setting_key.lower(),
                    getattr(
                        settings,
                        setting_key + self.s3_settings_suffix,
                        setting_default_value,
                    ),
                ),
            )
        # Validate settings.
        if not self.settings.AWS_S3_BUCKET_NAME:
            raise ImproperlyConfigured(
                f"Setting AWS_S3_BUCKET_NAME{self.s3_settings_suffix} is required."
            )
        # Create a thread-local connection manager.
        self._connections = _Local(self)
        # Set transfer config for S3 operations
        self._transfer_config = TransferConfig(
            use_threads=self.settings.AWS_S3_USE_THREADS
        )

    @property
    def s3_connection(self):
        return self._connections.s3_connection

    def _setting_changed_received(self, setting, **kwargs):
        if setting.startswith("AWS_"):
            self._setup()

    def __init__(self, **kwargs):
        # Check for unknown kwargs.
        for kwarg_key in kwargs.keys():
            if (
                kwarg_key.upper() not in self.default_auth_settings
                and kwarg_key.upper() not in self.default_s3_settings
            ):
                raise ImproperlyConfigured(f"Unknown S3Storage parameter: {kwarg_key}")
        # Set up the storage.
        self._kwargs = kwargs
        self._setup()
        # Re-initialize the storage if an AWS setting changes.
        setting_changed.connect(self._setting_changed_received)
        # Register system checks.
        checks.register(
            partial(self.__class__._system_checks, self), checks.Tags.security
        )
        # All done!
        super().__init__()

    def _system_checks(self, app_configs, **kwargs):
        errors = []
        if self.settings.AWS_S3_PUBLIC_URL and self.settings.AWS_S3_BUCKET_AUTH:
            errors.append(
                checks.Warning(
                    f"Using AWS_S3_BUCKET_AUTH{self.s3_settings_suffix} "
                    f"with AWS_S3_PUBLIC_URL{self.s3_settings_suffix}. "
                    "Private files on S3 may be inaccessible via the public URL. "
                    "See https://github.com/etianen/django-s3-storage/issues/114 ",
                    id="django_s3_storage.W001",
                )
            )
        return errors

    def __reduce__(self):
        return unpickle_helper, (self.__class__, self._kwargs)

    # Helpers.

    def _get_key_name(self, name):
        if name.startswith("/"):
            name = name[1:]
        return posixpath.normpath(
            posixpath.join(self.settings.AWS_S3_KEY_PREFIX, _to_posix_path(name))
        )

    def _object_params(self, name):
        params = {
            "Bucket": self.settings.AWS_S3_BUCKET_NAME,
            "Key": self._get_key_name(name),
        }
        return params

    def _object_put_params(self, name):
        # Set basic params.
        params = {
            "ACL": "private" if self.settings.AWS_S3_BUCKET_AUTH else "public-read",
            "CacheControl": "{privacy},max-age={max_age}".format(
                privacy="private" if self.settings.AWS_S3_BUCKET_AUTH else "public",
                max_age=self.settings.AWS_S3_MAX_AGE_SECONDS,
            ),
            "Metadata": {
                key: _callable_setting(value, name)
                for key, value in self.settings.AWS_S3_METADATA.items()
            },
            "StorageClass": "REDUCED_REDUNDANCY"
            if self.settings.AWS_S3_REDUCED_REDUNDANCY
            else "STANDARD",
        }
        params.update(self._object_params(name))
        # Set content disposition.
        content_disposition = _callable_setting(
            self.settings.AWS_S3_CONTENT_DISPOSITION, name
        )
        if content_disposition:
            params["ContentDisposition"] = content_disposition
        # Set content langauge.
        content_langauge = _callable_setting(
            self.settings.AWS_S3_CONTENT_LANGUAGE, name
        )
        if content_langauge:
            params["ContentLanguage"] = content_langauge
        # Set server-side encryption.
        if (
            self.settings.AWS_S3_ENCRYPT_KEY
        ):  # If this if False / None / empty then no encryption
            if isinstance(self.settings.AWS_S3_ENCRYPT_KEY, str):
                params["ServerSideEncryption"] = self.settings.AWS_S3_ENCRYPT_KEY
                if self.settings.AWS_S3_KMS_ENCRYPTION_KEY_ID:
                    params["SSEKMSKeyId"] = self.settings.AWS_S3_KMS_ENCRYPTION_KEY_ID
            else:
                params["ServerSideEncryption"] = "AES256"
        # All done!
        return params

    def new_temporary_file(self):
        """Returns a new file to use when opening from or saving to S3"""
        return SpooledTemporaryFile(max_size=1024 * 1024 * 10)  # 10 MB.

    @_wrap_errors
    def _open(self, name, mode="rb"):
        if mode not in ("rb", "rt", "r"):
            raise ValueError("S3 files can only be opened in read-only mode")
        # Load the key into a temporary file. It would be nice to stream the
        # content, but S3 doesn't support seeking, which is sometimes needed.
        obj = self.s3_connection.get_object(**self._object_params(name))
        content = self.new_temporary_file()
        shutil.copyfileobj(obj["Body"], content)
        content.seek(0)
        # Un-gzip if required.
        if obj.get("ContentEncoding") == "gzip":
            content = gzip.GzipFile(name, "rb", fileobj=content)
        # Decode text if required.
        if "b" not in mode:
            content = TextIOWrapper(content)
        # All done!
        return S3File(content, name, self)

    @_wrap_errors
    def _save(self, name, content):
        put_params = self._object_put_params(name)
        temp_files = []
        # The Django file storage API always rewinds the file before saving,
        # therefor so should we.
        content.seek(0)
        # Convert content to bytes.
        if isinstance(content.file, TextIOBase):
            temp_file = self.new_temporary_file()
            temp_files.append(temp_file)
            for chunk in content.chunks():
                temp_file.write(force_bytes(chunk))
            temp_file.seek(0)
            content = temp_file
        # Calculate the content type.
        content_type, _ = mimetypes.guess_type(name, strict=False)
        content_type = content_type or "application/octet-stream"
        put_params["ContentType"] = content_type
        # Calculate the content encoding.
        if self.settings.AWS_S3_GZIP:
            # Check if the content type is compressible.
            content_type_family, content_type_subtype = content_type.lower().split("/")
            content_type_subtype = content_type_subtype.split("+")[-1]
            if content_type_family == "text" or content_type_subtype in (
                "xml",
                "json",
                "html",
                "javascript",
            ):
                # Compress the content.
                temp_file = self.new_temporary_file()
                temp_files.append(temp_file)
                with closing(gzip.GzipFile(name, "wb", 9, temp_file)) as gzip_file:
                    shutil.copyfileobj(content, gzip_file)
                # Only use the compressed version if the zipped version is actually smaller!
                orig_size = content.tell()
                if temp_file.tell() < orig_size:
                    temp_file.seek(0)
                    content = temp_file
                    put_params["ContentEncoding"] = "gzip"
                    put_params["Metadata"][
                        _UNCOMPRESSED_SIZE_META_KEY
                    ] = f"{orig_size:d}"
                else:
                    content.seek(0)
        # Save the file.
        # HACK: Patch the file object to prevent `upload_fileobj` from closing it.
        # https://github.com/boto/boto3/issues/929
        original_close = content.close
        content.close = lambda: None
        try:
            self.s3_connection.upload_fileobj(
                content,
                put_params.pop("Bucket"),
                put_params.pop("Key"),
                ExtraArgs=put_params,
                Config=self._transfer_config,
            )
        finally:
            # Restore the original close method.
            content.close = original_close
        # Close all temp files.
        for temp_file in temp_files:
            temp_file.close()
        # All done!
        return name

    # Subsiduary storage methods.

    @_wrap_path_impl
    def get_valid_name(self, name):
        return super().get_valid_name(name)

    @_wrap_path_impl
    def get_available_name(self, name, max_length=None):
        if self.settings.AWS_S3_FILE_OVERWRITE:
            return _to_posix_path(name)
        return super().get_available_name(name, max_length=max_length)

    @_wrap_path_impl
    def generate_filename(self, filename):
        return super().generate_filename(filename)

    @_wrap_errors
    def meta(self, name):
        """Returns a dictionary of metadata associated with the key."""
        return self.s3_connection.head_object(**self._object_params(name))

    @_wrap_errors
    def delete(self, name):
        self.s3_connection.delete_object(**self._object_params(name))

    @_wrap_errors
    def copy(self, src_name, dst_name):
        self.s3_connection.copy_object(
            CopySource=self._object_params(src_name), **self._object_params(dst_name)
        )

    @_wrap_errors
    def rename(self, src_name, dst_name):
        self.copy(src_name, dst_name)
        self.s3_connection.delete_object(**self._object_params(src_name))

    def exists(self, name):
        name = _to_posix_path(name)
        if name.endswith("/"):
            # This looks like a directory, but on S3 directories are virtual, so we need to see if the key starts
            # with this prefix.
            try:
                results = self.s3_connection.list_objects_v2(
                    Bucket=self.settings.AWS_S3_BUCKET_NAME,
                    MaxKeys=1,
                    Prefix=self._get_key_name(name)
                    + "/",  # Add the slash again, since _get_key_name removes it.
                )
            except ClientError:
                return False
            else:
                return "Contents" in results
        # This may be a file or a directory. Check if getting the file metadata throws an error.
        try:
            self.meta(name)
        except OSError:
            # It's not a file, but it might be a directory. Check again that it's not a directory.
            return self.exists(name + "/")
        else:
            return True

    def listdir(self, path):
        path = self._get_key_name(path)
        path = "" if path == "." else path + "/"
        # Look through the paths, parsing out directories and paths.
        files = []
        dirs = []
        paginator = self.s3_connection.get_paginator("list_objects_v2")
        pages = paginator.paginate(
            Bucket=self.settings.AWS_S3_BUCKET_NAME,
            Delimiter="/",
            Prefix=path,
        )
        for page in pages:
            for entry in page.get("Contents", ()):
                files.append(posixpath.relpath(entry["Key"], path))
            for entry in page.get("CommonPrefixes", ()):
                dirs.append(posixpath.relpath(entry["Prefix"], path))
        # All done!
        return dirs, files

    def size(self, name):
        meta = self.meta(name)
        try:
            if meta["ContentEncoding"] == "gzip":
                return int(meta["Metadata"]["uncompressed_size"])
        except KeyError:
            return meta["ContentLength"]

    def url(self, name, extra_params=None, client_method="get_object"):
        # Use a public URL, if specified.
        if self.settings.AWS_S3_PUBLIC_URL:
            if extra_params or client_method != "get_object":
                raise ValueError(
                    "Use of extra_params or client_method is not allowed with AWS_S3_PUBLIC_URL"
                )
            return urljoin(self.settings.AWS_S3_PUBLIC_URL, filepath_to_uri(name))
        # Otherwise, generate the URL.
        params = extra_params.copy() if extra_params else {}
        params.update(self._object_params(name))
        url = self.s3_connection.generate_presigned_url(
            ClientMethod=client_method,
            Params=params,
            ExpiresIn=self.settings.AWS_S3_MAX_AGE_SECONDS,
        )
        # Strip off the query params if we're not interested in bucket auth.
        if not self.settings.AWS_S3_BUCKET_AUTH:
            if extra_params or client_method != "get_object":
                raise ValueError(
                    "Use of extra_params or client_method is not allowed with AWS_S3_BUCKET_AUTH"
                )
            url = urlunsplit(
                urlsplit(url)[:3]
                + (
                    "",
                    "",
                )
            )
        # All done!
        return url

    def modified_time(self, name):
        return make_naive(self.meta(name)["LastModified"], timezone.utc)

    created_time = accessed_time = modified_time

    def get_modified_time(self, name):
        timestamp = self.meta(name)["LastModified"]
        return timestamp if settings.USE_TZ else make_naive(timestamp)

    get_created_time = get_accessed_time = get_modified_time

    def sync_meta_iter(self):
        paginator = self.s3_connection.get_paginator("list_objects_v2")
        pages = paginator.paginate(
            Bucket=self.settings.AWS_S3_BUCKET_NAME,
            Prefix=self.settings.AWS_S3_KEY_PREFIX,
        )
        for page in pages:
            for entry in page.get("Contents", ()):
                name = posixpath.relpath(entry["Key"], self.settings.AWS_S3_KEY_PREFIX)
                try:
                    obj = self.meta(name)
                except OSError:
                    # This may be caused by a race condition, with the entry being deleted before it was accessed.
                    # Alternatively, the key may be something that, when normalized, has a different path, which will
                    # mean that the key's meta cannot be accessed.
                    continue
                put_params = self._object_put_params(name)
                # Set content encoding.
                content_encoding = obj.get("ContentEncoding")
                if content_encoding:
                    put_params["ContentEncoding"] = content_encoding
                    if content_encoding == "gzip":
                        try:
                            put_params["Metadata"][_UNCOMPRESSED_SIZE_META_KEY] = obj[
                                "Metadata"
                            ][_UNCOMPRESSED_SIZE_META_KEY]
                        except KeyError:
                            pass
                # Update the metadata.
                self.s3_connection.copy_object(
                    ContentType=obj["ContentType"],
                    CopySource={
                        "Bucket": self.settings.AWS_S3_BUCKET_NAME,
                        "Key": self._get_key_name(name),
                    },
                    MetadataDirective="REPLACE",
                    **put_params,
                )
                yield name

    def sync_meta(self):
        for path in self.sync_meta_iter():
            pass


class StaticS3Storage(S3Storage):

    """
    An S3 storage for storing static files.
    """

    default_s3_settings = S3Storage.default_s3_settings.copy()
    default_s3_settings.update(
        {
            "AWS_S3_BUCKET_AUTH": False,
        }
    )

    s3_settings_suffix = "_STATIC"


class ManifestStaticS3Storage(ManifestFilesMixin, StaticS3Storage):
    default_s3_settings = StaticS3Storage.default_s3_settings.copy()
    default_s3_settings.update(
        {
            "AWS_S3_MAX_AGE_SECONDS_CACHED": 60 * 60 * 24 * 365,  # 1 year.
        }
    )

    def post_process(self, *args, **kwargs):
        initial_aws_s3_max_age_seconds = self.settings.AWS_S3_MAX_AGE_SECONDS
        self.settings.AWS_S3_MAX_AGE_SECONDS = (
            self.settings.AWS_S3_MAX_AGE_SECONDS_CACHED
        )
        try:
            yield from super().post_process(*args, **kwargs)
        finally:
            self.settings.AWS_S3_MAX_AGE_SECONDS = initial_aws_s3_max_age_seconds
